File: src/main/java/io/github/takenoha/towerdefense/domain/TeamProgress.java
package io.github.takenoha.towerdefense.domain;

import java.util.Objects;
import java.util.UUID;

/** Durable team-wide progression values used by the economy and future research slices. */
public record TeamProgress(
        UUID teamId,
        long highestClearedLevel,
        long unlockedLevel,
        long researchPoints) {
    public TeamProgress {
        Objects.requireNonNull(teamId, "teamId");
        if (highestClearedLevel < 0L) {
            throw new IllegalArgumentException("highestClearedLevel must be non-negative");
        }
        if (unlockedLevel <= 0L) {
            throw new IllegalArgumentException("unlockedLevel must be positive");
        }
        if (researchPoints < 0L) {
            throw new IllegalArgumentException("researchPoints must be non-negative");
        }
        long minimumUnlock = highestClearedLevel >= StageWaveSchedule.MAX_STAGE_LEVEL
                ? StageWaveSchedule.MAX_STAGE_LEVEL
                : highestClearedLevel + 1L;
        if (unlockedLevel < minimumUnlock) {
            throw new IllegalArgumentException(
                    "unlockedLevel must include the next level after the highest clear");
        }
    }

    public static TeamProgress initial(UUID teamId) {
        return new TeamProgress(teamId, 0L, 1L, 0L);
    }

    /**
     * Returns the monotonic progression snapshot produced by a stage victory.
     *
     * <p>Defeat and technical recovery deliberately do not call this method. The next-stage
     * unlock is derived from the highest clear and is therefore safe to replay inside the
     * terminal transaction.</p>
     */
    public TeamProgress afterVictory(long stageLevel) {
        if (stageLevel <= 0L) {
            throw new IllegalArgumentException("stageLevel must be positive");
        }
        long nextUnlocked = StageWaveSchedule.nextStageLevel(stageLevel)
                .orElse(stageLevel);
        return new TeamProgress(
                teamId,
                Math.max(highestClearedLevel, stageLevel),
                Math.max(unlockedLevel, nextUnlocked),
                researchPoints);
    }
}

File: src/main/kotlin/io/github/takenoha/towerdefense/domain/TeamProgress.kt
package io.github.takenoha.towerdefense.domain

import java.util.Objects
import java.util.UUID
import kotlin.jvm.JvmRecord

/** Durable team-wide progression values used by the economy and future research slices. */
@JvmRecord
data class TeamProgress(
    val teamId: UUID,
    val highestClearedLevel: Long,
    val unlockedLevel: Long,
    val researchPoints: Long,
) {
    init {
        Objects.requireNonNull(teamId, "teamId")
        if (highestClearedLevel < 0L) {
            throw IllegalArgumentException("highestClearedLevel must be non-negative")
        }
        if (unlockedLevel <= 0L) {
            throw IllegalArgumentException("unlockedLevel must be positive")
        }
        if (researchPoints < 0L) {
            throw IllegalArgumentException("researchPoints must be non-negative")
        }
        val minimumUnlock = if (highestClearedLevel >= StageWaveSchedule.MAX_STAGE_LEVEL) {
            StageWaveSchedule.MAX_STAGE_LEVEL
        } else {
            highestClearedLevel + 1L
        }
        if (unlockedLevel < minimumUnlock) {
            throw IllegalArgumentException(
                "unlockedLevel must include the next level after the highest clear",
            )
        }
    }

    companion object {
        @JvmStatic
        fun initial(teamId: UUID): TeamProgress = TeamProgress(teamId, 0L, 1L, 0L)
    }

    /** Returns the monotonic progression snapshot produced by a stage victory. */
    fun afterVictory(stageLevel: Long): TeamProgress {
        if (stageLevel <= 0L) {
            throw IllegalArgumentException("stageLevel must be positive")
        }
        val nextUnlocked = StageWaveSchedule.nextStageLevel(stageLevel)
            .orElse(stageLevel)
        return TeamProgress(
            teamId,
            maxOf(highestClearedLevel, stageLevel),
            maxOf(unlockedLevel, nextUnlocked),
            researchPoints,
        )
    }
}
